File: src/main/java/top/mcfpp/nbt/tags/primitive/ByteTag.java
package top.mcfpp.nbt.tags.primitive;

import top.mcfpp.nbt.visitors.StringTagVisitor;
import top.mcfpp.nbt.visitors.TagVisitor;

public record ByteTag(byte value) implements NumericTag {
	public static final ByteTag ZERO = valueOf((byte)0);
	public static final ByteTag ONE = valueOf((byte)1);

	public static ByteTag valueOf(byte b) {
		return Cache.cache[128 + b];
	}

	public static ByteTag valueOf(boolean bl) {
		return bl ? ONE : ZERO;
	}

	public ByteTag copy() {
		return this;
	}

	@Override
	public void accept(TagVisitor tagVisitor) {
		tagVisitor.visitByte(this);
	}

	@Override
	public long longValue() {
		return this.value;
	}

	@Override
	public int intValue() {
		return this.value;
	}

	@Override
	public short shortValue() {
		return this.value;
	}

	@Override
	public byte byteValue() {
		return this.value;
	}

	@Override
	public double doubleValue() {
		return this.value;
	}

	@Override
	public float floatValue() {
		return this.value;
	}

	@Override
	public Number box() {
		return this.value;
	}

	@Override
	public String toString() {
		StringTagVisitor stringTagVisitor = new StringTagVisitor();
		stringTagVisitor.visitByte(this);
		return stringTagVisitor.build();
	}

	static class Cache {
		static final ByteTag[] cache = new ByteTag[256];

		private Cache() {
		}

		static {
			for (int i = 0; i < cache.length; i++) {
				cache[i] = new ByteTag((byte)(i - 128));
			}
		}
	}
}

File: src/main/kotlin/top/mcfpp/nbt/tags/primitive/ByteTag.kt
package top.mcfpp.nbt.tags.primitive

import top.mcfpp.nbt.visitors.StringTagVisitor
import top.mcfpp.nbt.visitors.TagVisitor

@JvmRecord
data class ByteTag(override val value: Byte) : NumericTag<Byte> {

    constructor() : this(0.toByte())

    constructor(value: Boolean) : this(if(value) 1.toByte() else 0.toByte())

    override fun copy(): ByteTag {
        return this
    }

    override fun accept(tagVisitor: TagVisitor) {
        tagVisitor.visitByte(this)
    }

    override fun longValue(): Long {
        return value.toLong()
    }

    override fun intValue(): Int {
        return value.toInt()
    }

    override fun shortValue(): Short {
        return value.toShort()
    }

    override fun byteValue(): Byte {
        return this.value
    }

    override fun doubleValue(): Double {
        return value.toDouble()
    }

    override fun floatValue(): Float {
        return value.toFloat()
    }

    override fun box(): Number {
        return this.value
    }

    override fun toString(): String {
        val stringTagVisitor = StringTagVisitor()
        stringTagVisitor.visitByte(this)
        return stringTagVisitor.build()
    }

    internal object Cache {
        val cache: Array<ByteTag> = Array(256){
            ByteTag((it - 128).toByte())
        }
    }

    companion object {
        val ZERO: ByteTag = valueOf(0.toByte())
        val ONE: ByteTag = valueOf(1.toByte())

        @JvmStatic
		fun valueOf(b: Byte): ByteTag {
            return Cache.cache[128 + b]
        }

        @JvmStatic
		fun valueOf(bl: Boolean): ByteTag {
            return if (bl) ONE else ZERO
        }
    }
}
